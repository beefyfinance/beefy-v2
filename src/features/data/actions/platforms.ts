import { createAsyncThunk } from '@reduxjs/toolkit';
import { getConfigApi } from '../apis/instances';
import { PlatformConfig } from '../apis/config-types';

export type FetchPlatformsPayload = PlatformConfig[];

export const fetchPlatforms = createAsyncThunk<FetchPlatformsPayload>(
  'platforms/fetchPlatforms',
  async () => {
    const api = getConfigApi();
    return await api.fetchPlatforms();
  }
);
