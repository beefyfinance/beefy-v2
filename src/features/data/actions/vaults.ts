import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import { getBeefyApi, getConfigApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import type { FeaturedVaultConfig, VaultConfig } from '../apis/config-types';
import type { BeefyVaultZapSupportResponse } from '../apis/beefy/beefy-api';
import { featureFlag_zapSupportOverrides } from '../utils/feature-flags';

// given the list of vaults is pulled from some api at some point
// we use the api to create an action
// this action should return just enough data for the state to work with

export interface FulfilledAllVaultsPayload {
  byChainId: {
    [chainId: ChainEntity['id']]: VaultConfig[];
  };
  state: BeefyState;
}
export const fetchAllVaults = createAsyncThunk<
  FulfilledAllVaultsPayload,
  void,
  { state: BeefyState }
>('vaults/fetchAllVaults', async (_, { getState }) => {
  const api = getConfigApi();
  const vaults = await api.fetchAllVaults();
  return { byChainId: vaults, state: getState() };
});

export interface FulfilledFeaturedVaultsPayload {
  byVaultId: FeaturedVaultConfig;
}
export const fetchFeaturedVaults = createAsyncThunk<FulfilledFeaturedVaultsPayload>(
  'vaults/fetchFeaturedVaults',
  async () => {
    const api = getConfigApi();
    const featuredVaults = await api.fetchFeaturedVaults();
    return { byVaultId: featuredVaults };
  }
);

export type FulfilledFetchVaultsZapSupportPayload = {
  byVaultId: BeefyVaultZapSupportResponse;
};

export const fetchVaultsZapSupport = createAsyncThunk<
  FulfilledFetchVaultsZapSupportPayload,
  void,
  { state: BeefyState }
>('vaults/fetchVaultsZapSupport', async (_, { getState }) => {
  const api = getBeefyApi();
  const vaultsZapSupport = await api.getVaultZapSupport();
  const overrides = featureFlag_zapSupportOverrides();

  for (const kind of ['beefy', 'oneInch'] as const) {
    const value = overrides[kind];
    let ids = [];
    if (value === 'all') {
      ids = getState().entities.vaults.allIds;
    } else if (Array.isArray(value) && value.length > 0) {
      ids = value;
    }
    for (const id of ids) {
      vaultsZapSupport[id] = [...(vaultsZapSupport[id] || []), kind];
    }
  }

  return { byVaultId: vaultsZapSupport };
});

type FulfilledVaultsLastHarvestPayload = {
  byVaultId: { [vaultId: VaultConfig['id']]: number };
};

export const fetchVaultsLastHarvests = createAsyncThunk<FulfilledVaultsLastHarvestPayload>(
  'vaults/last-harvest',
  async () => {
    const api = getBeefyApi();
    const vaults = await api.getVaultLastHarvest();
    return { byVaultId: vaults };
  }
);
