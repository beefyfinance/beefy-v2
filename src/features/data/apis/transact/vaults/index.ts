import type { VaultType, VaultTypeConstructor } from './IVaultType';
import type { VaultEntity } from '../../../entities/vault';
import type { GetStateFn } from '../../../../../redux-types';

function makeLazyLoader<T extends VaultType>(loader: () => Promise<VaultTypeConstructor<T>>) {
  let constructor: VaultTypeConstructor<T> | undefined;

  return async (vault: VaultEntity, getState: GetStateFn) => {
    if (!constructor) {
      constructor = await loader();
    }

    return new constructor(vault, getState);
  };
}

export const vaultTypeBuildersById = {
  gov: makeLazyLoader(async () => (await import('./GovVaultType')).GovVaultType),
  standard: makeLazyLoader(async () => (await import('./StandardVaultType')).StandardVaultType),
  cowcentrated: makeLazyLoader(
    async () => (await import('./CowcentratedVaultType')).CowcentratedVaultType
  ),
} as const satisfies Record<
  VaultEntity['type'],
  (vault: VaultEntity, getState: GetStateFn) => Promise<VaultType>
>;
