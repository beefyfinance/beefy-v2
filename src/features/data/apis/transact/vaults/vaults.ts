import type { VaultTypeConstructor } from './IVaultType.ts';
import type { VaultEntity } from '../../../entities/vault.ts';
import type { BeefyStateFn } from '../../../../../redux-types.ts';

function makeLazyLoader<T extends VaultEntity>(loader: () => Promise<VaultTypeConstructor<T>>) {
  let constructor: VaultTypeConstructor<T> | undefined;

  return async (vault: T, getState: BeefyStateFn) => {
    if (!constructor) {
      constructor = await loader();
    }

    return new constructor(vault, getState);
  };
}

type VaultTypeBuilderFromVault<T extends VaultEntity> = ReturnType<typeof makeLazyLoader<T>>;

type TypeToConstructorMap = {
  [K in VaultEntity['type']]: VaultTypeBuilderFromVault<
    Extract<
      VaultEntity,
      {
        type: K;
      }
    >
  >;
};

const vaultTypeBuildersById: TypeToConstructorMap = {
  gov: makeLazyLoader(async () => (await import('./GovVaultType.ts')).GovVaultType),
  standard: makeLazyLoader(async () => (await import('./StandardVaultType.ts')).StandardVaultType),
  cowcentrated: makeLazyLoader(
    async () => (await import('./CowcentratedVaultType.ts')).CowcentratedVaultType
  ),
  erc4626: makeLazyLoader(async () => (await import('./Erc4626VaultType.ts')).Erc4626VaultType),
};

export function getVaultTypeBuilder<T extends VaultEntity>(vault: T): VaultTypeBuilderFromVault<T> {
  return vaultTypeBuildersById[vault.type] as unknown as VaultTypeBuilderFromVault<T>; // ???
}
