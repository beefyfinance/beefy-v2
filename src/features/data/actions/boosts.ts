import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { TokenAllowance } from '../apis/allowance/allowance-types';
import type { FetchAllBalancesResult } from '../apis/balance/balance-types';
import { getAllowanceApi, getBalanceApi, getConfigApi } from '../apis/instances';
import type { BoostEntity } from '../entities/boost';
import type { ChainEntity } from '../entities/chain';
import { selectBoostById } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectErc20TokenByAddress } from '../selectors/tokens';
import { selectVaultById } from '../selectors/vaults';
import type { BoostCampaignConfig, BoostConfig, BoostPartnerConfig } from '../apis/config-types';

// given the list of vaults is pulled from some api at some point
// we use the api to create an action
// this action should return just enough data for the state to work with

export interface FulfilledAllBoostsPayload {
  boostsByChainId: Record<ChainEntity['id'], BoostConfig[]>;
  partnersById: Record<string, BoostPartnerConfig>;
  campaignsById: Record<string, BoostCampaignConfig>;
}

export const fetchAllBoosts = createAsyncThunk<FulfilledAllBoostsPayload>(
  'boosts/fetchAllBoosts',
  async () => {
    const api = await getConfigApi();
    return api.fetchAllBoosts();
  }
);

interface InitBoostFormParams {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';
  walletAddress: string | undefined;
}

interface InitBoostFormPayload {
  boostId: BoostEntity['id'];
  mode: 'stake' | 'unstake';

  // really, this should be separated
  walletAddress: string | undefined;
  balance: FetchAllBalancesResult;
  allowance: TokenAllowance[];

  // reducers below need to access the state
  state: BeefyState;
}

export const initiateBoostForm = createAsyncThunk<
  InitBoostFormPayload,
  InitBoostFormParams,
  { state: BeefyState }
>('boosts/initBoostForm', async ({ boostId, mode, walletAddress }, { getState }) => {
  const boost = selectBoostById(getState(), boostId);
  const vault = selectVaultById(getState(), boost.vaultId);
  const chain = selectChainById(getState(), boost.chainId);

  const balanceApi = await getBalanceApi(chain);

  const balanceRes: FetchAllBalancesResult = walletAddress
    ? await balanceApi.fetchAllBalances(getState(), [], [], [boost], walletAddress)
    : { tokens: [], boosts: [], govVaults: [] };

  const spenderAddress = boost.earnContractAddress;

  const allowanceApi = await getAllowanceApi(chain);
  const mooToken = selectErc20TokenByAddress(
    getState(),
    boost.chainId,
    vault.earnedTokenAddress,
    false
  );
  const allowanceRes =
    walletAddress && spenderAddress
      ? await allowanceApi.fetchTokensAllowance(
          getState(),
          [mooToken],
          walletAddress,
          spenderAddress
        )
      : [];

  return {
    boostId,
    mode,
    walletAddress,
    allowance: allowanceRes,
    balance: balanceRes,
    state: getState(),
  };
});
