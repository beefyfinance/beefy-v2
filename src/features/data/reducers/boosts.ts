import { createSlice } from '@reduxjs/toolkit';
import { WritableDraft } from 'immer/dist/internal';
import { fetchAllBoosts } from '../actions/boosts';
import { BoostConfig } from '../apis/config';
import { BoostEntity, isBoostActive, isBoostPreStake } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { VaultEntity } from '../entities/vault';
import { NormalizedEntity } from '../utils/normalized-entity';

/**
 * State containing Vault infos
 */
export type BoostsState = NormalizedEntity<BoostEntity> & {
  byVaultId: {
    [vaultId: VaultEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
  byChainId: {
    [chainId: ChainEntity['id']]: {
      allBoostsIds: BoostEntity['id'][];
      prestakeBoostsIds: BoostEntity['id'][];
      activeBoostsIds: BoostEntity['id'][];
    };
  };
};
export const initialBoostsState: BoostsState = {
  byId: {},
  allIds: [],
  byVaultId: {},
  byChainId: {},
};

export const boostsSlice = createSlice({
  name: 'boosts',
  initialState: initialBoostsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when boost list is fetched, add all new tokens
    builder.addCase(fetchAllBoosts.fulfilled, (sliceState, action) => {
      for (const [chainId, boosts] of Object.entries(action.payload)) {
        for (const boost of boosts) {
          addBoostToState(sliceState, chainId, boost);
        }
      }
    });
  },
});

function addBoostToState(
  sliceState: WritableDraft<BoostsState>,
  chainId: ChainEntity['id'],
  apiBoost: BoostConfig
) {
  if (apiBoost.id in sliceState.byId) {
    return;
  }

  const boost: BoostEntity = {
    id: apiBoost.id,
    chainId: chainId,
    assets: apiBoost.assets,
    earnedTokenId: apiBoost.earnedOracleId,
    earnContractAddress: apiBoost.earnContractAddress,
    logo: apiBoost.logo,
    name: apiBoost.name,
    partnerIds: apiBoost.partners ? apiBoost.partners.map(p => p.website) : [],
    status: apiBoost.status as BoostEntity['status'],
    vaultId: apiBoost.poolId,
  };
  sliceState.byId[boost.id] = boost;
  sliceState.allIds.push(boost.id);

  // add to vault id index
  if (sliceState.byVaultId[boost.vaultId] === undefined) {
    sliceState.byVaultId[boost.vaultId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byVaultId[boost.vaultId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byVaultId[boost.vaultId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byVaultId[boost.vaultId].prestakeBoostsIds.push(boost.id);
  }

  // add to chain id index
  if (sliceState.byChainId[chainId] === undefined) {
    sliceState.byChainId[chainId] = {
      allBoostsIds: [],
      activeBoostsIds: [],
      prestakeBoostsIds: [],
    };
  }
  sliceState.byChainId[chainId].allBoostsIds.push(boost.id);
  if (isBoostActive(boost)) {
    sliceState.byChainId[chainId].activeBoostsIds.push(boost.id);
  }
  if (isBoostPreStake(boost)) {
    sliceState.byChainId[chainId].prestakeBoostsIds.push(boost.id);
  }
}
