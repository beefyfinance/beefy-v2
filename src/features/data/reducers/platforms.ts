import { createSlice } from '@reduxjs/toolkit';
import type { Draft } from 'immer';
import type { PlatformEntity } from '../entities/platform';
import type { NormalizedEntity } from '../utils/normalized-entity';
import { fetchPlatforms } from '../actions/platforms';
import { fetchAllVaults } from '../actions/vaults';
import type { PlatformConfig, VaultConfig } from '../apis/config-types';

/**
 * State containing Vault infos
 */
export type PlatformsState = NormalizedEntity<PlatformEntity> & {
  filterIds: PlatformEntity['id'][];
  activeIds: PlatformEntity['id'][];
};
export const initialPlatformsState: PlatformsState = {
  byId: {},
  allIds: [],
  filterIds: [],
  activeIds: [],
};

export const platformsSlice = createSlice({
  name: 'platforms',
  initialState: initialPlatformsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    // when vault list is fetched, add all new tokens
    builder.addCase(fetchAllVaults.fulfilled, (sliceState, action) => {
      for (const vaults of Object.values(action.payload.byChainId)) {
        for (const vault of vaults) {
          addVaultToState(sliceState, vault);
        }
      }
    });

    builder.addCase(fetchPlatforms.fulfilled, (sliceState, action) => {
      for (const platform of action.payload) {
        addPlatformToState(sliceState, platform);
      }
    });
  },
});

function addVaultToState(sliceState: Draft<PlatformsState>, vault: VaultConfig) {
  if (vault.status !== 'eol') {
    if (vault.platformId && !sliceState.activeIds.includes(vault.platformId)) {
      sliceState.activeIds.push(vault.platformId);
    }
    if (vault.tokenProviderId && !sliceState.activeIds.includes(vault.tokenProviderId)) {
      sliceState.activeIds.push(vault.tokenProviderId);
    }
  }
}

function addPlatformToState(sliceState: Draft<PlatformsState>, platformConfig: PlatformConfig) {
  if (sliceState.byId[platformConfig.id] === undefined) {
    const platform: PlatformEntity = {
      id: platformConfig.id,
      name: platformConfig.name,
    };
    sliceState.byId[platform.id] = platform;
    sliceState.allIds.push(platform.id);

    // keep list of filter platforms
    if (platformConfig.filter) {
      sliceState.filterIds.push(platform.id);
    }
  }
}
