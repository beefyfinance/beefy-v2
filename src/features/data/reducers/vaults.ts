import { createSlice } from '@reduxjs/toolkit';
import type BigNumber from 'bignumber.js';
import type { Draft } from 'immer';
import { isEmpty, mapValues, sortBy } from 'lodash-es';
import { safetyScoreNum } from '../../../helpers/safetyScore';
import type { BeefyState } from '../../../redux-types';
import { fetchAllContractDataByChainAction } from '../actions/contract-data';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens';
import {
  fetchAllVaults,
  fetchFeaturedVaults,
  fetchVaultsZapSupport,
  fetchVaultsLastHarvests,
} from '../actions/vaults';
import type { FetchAllContractDataResult } from '../apis/contract-data/contract-data-types';
import type { ChainEntity } from '../entities/chain';
import type { VaultEntity, VaultGov, VaultStandard } from '../entities/vault';
import type { NormalizedEntity } from '../utils/normalized-entity';
import type { FeaturedVaultConfig, VaultConfig } from '../apis/config-types';

/**
 * State containing Vault infos
 */
export type VaultsState = NormalizedEntity<VaultEntity> & {
  byChainId: {
    [chainId: ChainEntity['id']]: {
      // add quick access arrays
      // doesn't need to be by chain but it's convenient
      // for when we load by chain
      allActiveIds: VaultEntity['id'][];
      allRetiredIds: VaultEntity['id'][];

      // used to find a vault by it's token for balance stuff
      standardVault: {
        byDepositTokenAddress: {
          [address: string]: VaultEntity['id'][];
        };
        byEarnedTokenAddress: {
          [address: string]: VaultEntity['id'];
        };
      };
      govVault: {
        byDepositTokenAddress: {
          [address: string]: VaultEntity['id'][];
        };
      };
    };
  };

  /**
   * pricePerFullShare is how you find out how much your mooTokens
   * (shares) represent in term of the underlying asset
   *
   * So if you deposit 1 BIFI you will get, for example 0.95 mooBIFI,
   * with a ppfs of X, if you multiply your mooBIIFI * ppfs you get your amount in BIFI
   *
   * That value is fetched from the smart contract upon loading
   **/
  contractData: {
    byVaultId: {
      [vaultId: VaultEntity['id']]: {
        strategyAddress: string;
        pricePerFullShare: BigNumber;
      };
    };
  };

  /**
   * We want to know if the vault is featured or not
   */
  featuredVaults: FeaturedVaultConfig;

  /**
   * Which zaps are enabled for each vault
   */
  zapSupportById: {
    [vaultId: VaultEntity['id']]: {
      beefy: boolean;
      oneInch: boolean;
    };
  };

  lastHarvestById: {
    [vaultId: VaultEntity['id']]: number;
  };
};
export const initialVaultsState: VaultsState = {
  byId: {},
  allIds: [],
  byChainId: {},
  contractData: { byVaultId: {} },
  featuredVaults: {},
  zapSupportById: {},
  lastHarvestById: {},
};

export const vaultsSlice = createSlice({
  name: 'vaults',
  initialState: initialVaultsState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchFeaturedVaults.fulfilled, (sliceState, action) => {
      sliceState.featuredVaults = action.payload.byVaultId;
    });

    builder.addCase(fetchAllVaults.fulfilled, (sliceState, action) => {
      const initialVaultAmount = sliceState.allIds.length;
      for (const [chainId, vaults] of Object.entries(action.payload.byChainId)) {
        for (const vault of vaults) {
          addVaultToState(action.payload.state, sliceState, chainId, vault);
        }
      }

      // If new vaults were added, apply default sorting
      if (sliceState.allIds.length !== initialVaultAmount) {
        sliceState.allIds = sortBy(sliceState.allIds, id => {
          return -sliceState.byId[id].createdAt;
        });
      }
    });

    builder.addCase(fetchAllContractDataByChainAction.fulfilled, (sliceState, action) => {
      addContractDataToState(sliceState, action.payload.data);
    });

    builder.addCase(fetchVaultsLastHarvests.fulfilled, (sliceState, action) => {
      for (const [vaultId, lastHarvest] of Object.entries(action.payload.byVaultId)) {
        // time in milliseconds since unix epoch
        sliceState.lastHarvestById[vaultId] = lastHarvest * 1000;
      }
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        addContractDataToState(sliceState, action.payload.contractData);
      }
    );

    builder.addCase(fetchVaultsZapSupport.fulfilled, (sliceState, action) => {
      sliceState.zapSupportById = mapValues(action.payload.byVaultId, support => {
        return {
          beefy: support.includes('beefy'),
          oneInch: support.includes('oneInch'),
        };
      });
    });
  },
});

function addContractDataToState(
  sliceState: Draft<VaultsState>,
  contractData: FetchAllContractDataResult
) {
  for (const vaultContractData of contractData.standardVaults) {
    const vaultId = vaultContractData.id;

    // only update it if needed
    if (sliceState.contractData.byVaultId[vaultId] === undefined) {
      sliceState.contractData.byVaultId[vaultId] = {
        pricePerFullShare: vaultContractData.pricePerFullShare,
        strategyAddress: vaultContractData.strategy,
      };
    }

    if (
      !sliceState.contractData.byVaultId[vaultId].pricePerFullShare.isEqualTo(
        vaultContractData.pricePerFullShare
      )
    ) {
      sliceState.contractData.byVaultId[vaultId].pricePerFullShare =
        vaultContractData.pricePerFullShare;
    }
    if (sliceState.contractData.byVaultId[vaultId].strategyAddress !== vaultContractData.strategy) {
      sliceState.contractData.byVaultId[vaultId].strategyAddress = vaultContractData.strategy;
    }
  }
}

function addVaultToState(
  state: BeefyState,
  sliceState: Draft<VaultsState>,
  chainId: ChainEntity['id'],
  apiVault: VaultConfig
) {
  // we already know this vault
  if (apiVault.id in sliceState.byId) {
    return;
  }
  const score = getVaultSafetyScore(state, chainId, apiVault);

  if (apiVault.isGovVault) {
    const vault: VaultGov = {
      id: apiVault.id,
      name: apiVault.name,
      isGovVault: true,
      depositTokenAddress: apiVault.tokenAddress,
      earnedTokenAddress: apiVault.earnedTokenAddress,
      earnContractAddress: apiVault.earnContractAddress,
      excludedId: apiVault.excluded || null,
      chainId: chainId,
      status: apiVault.status as VaultGov['status'],
      platformId: apiVault.platformId,
      safetyScore: score,
      assetIds: apiVault.assets || [],
      type: 'single',
      risks: apiVault.risks || [],
      buyTokenUrl: apiVault.buyTokenUrl || null,
      addLiquidityUrl: null,
      removeLiquidityUrl: null,
      depositFee: apiVault.depositFee ?? '0%',
      createdAt: apiVault.createdAt ?? 0,
      retireReason: apiVault.retireReason,
      pauseReason: apiVault.pauseReason,
    };

    sliceState.byId[vault.id] = vault;
    sliceState.allIds.push(vault.id);
    if (sliceState.byChainId[vault.chainId] === undefined) {
      sliceState.byChainId[vault.chainId] = {
        allActiveIds: [],
        allRetiredIds: [],
        standardVault: {
          byEarnedTokenAddress: {},
          byDepositTokenAddress: {},
        },
        govVault: {
          byDepositTokenAddress: {},
        },
      };
    }

    const vaultState = sliceState.byChainId[vault.chainId];
    if (apiVault.status === 'eol' || apiVault.status === 'paused') {
      vaultState.allRetiredIds.push(vault.id);
    } else {
      vaultState.allActiveIds.push(vault.id);
    }

    if (!vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()]) {
      vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()] = [];
    }
    vaultState.govVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()].push(
      vault.id
    );
  } else {
    const vault: VaultStandard = {
      id: apiVault.id,
      name: apiVault.name,
      isGovVault: false,
      depositTokenAddress: apiVault.tokenAddress ?? 'native',
      earnContractAddress: apiVault.earnContractAddress,
      earnedTokenAddress: apiVault.earnedTokenAddress,
      strategyTypeId: apiVault.strategyTypeId,
      chainId: chainId,
      platformId: apiVault.platformId,
      status: apiVault.status as VaultStandard['status'],
      type: !apiVault.assets ? 'single' : apiVault.assets.length > 1 ? 'lps' : 'single',
      safetyScore: score,
      assetIds: apiVault.assets || [],
      risks: apiVault.risks || [],
      buyTokenUrl: apiVault.buyTokenUrl || null,
      addLiquidityUrl: apiVault.addLiquidityUrl || null,
      removeLiquidityUrl: apiVault.removeLiquidityUrl || null,
      depositFee: apiVault.depositFee ?? '0%',
      createdAt: apiVault.createdAt ?? 0,
      retireReason: apiVault.retireReason,
      pauseReason: apiVault.pauseReason,
      migrationIds: apiVault.migrationIds,
    };
    // redux toolkit uses immer by default so we can
    // directly modify the state as usual
    sliceState.byId[vault.id] = vault;
    sliceState.allIds.push(vault.id);
    if (sliceState.byChainId[vault.chainId] === undefined) {
      sliceState.byChainId[vault.chainId] = {
        allActiveIds: [],
        allRetiredIds: [],
        standardVault: {
          byEarnedTokenAddress: {},
          byDepositTokenAddress: {},
        },
        govVault: {
          byDepositTokenAddress: {},
        },
      };
    }
    const vaultState = sliceState.byChainId[vault.chainId];
    if (apiVault.status === 'eol' || apiVault.status === 'paused') {
      vaultState.allRetiredIds.push(vault.id);
    } else {
      vaultState.allActiveIds.push(vault.id);
    }

    if (!vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()]) {
      vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()] = [];
    }
    vaultState.standardVault.byDepositTokenAddress[vault.depositTokenAddress.toLowerCase()].push(
      vault.id
    );
    vaultState.standardVault.byEarnedTokenAddress[vault.earnedTokenAddress.toLowerCase()] =
      vault.id;
  }
}

function getVaultSafetyScore(
  state: BeefyState,
  chainId: ChainEntity['id'],
  apiVault: VaultConfig
): number {
  let score = 0;
  if (!isEmpty(apiVault.risks)) {
    score = safetyScoreNum(apiVault.risks);
  }

  return score;
}
