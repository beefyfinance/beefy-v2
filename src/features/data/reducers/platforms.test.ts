import type { BeefyState } from '../../../redux-types';
import type { FulfilledAllBoostsPayload } from '../actions/boosts';
import type { FulfilledAllVaultsPayload } from '../actions/vaults';
import { fetchAllBoosts } from '../actions/boosts';
import { fetchAllVaults } from '../actions/vaults';
import { describe, expect, it } from 'vitest';
import { platformsSlice, initialPlatformsState } from './platforms';

describe('Platforms slice tests', () => {
  it('should update state on fulfilled vault list', () => {
    const payload: FulfilledAllVaultsPayload = {
      state: {} as BeefyState,
      byChainId: {
        harmony: [
          // have one gov vault
          {
            id: 'one-bifi-gov',
            name: 'BIFI Earnings Pool',
            token: 'BIFI',
            tokenAddress: '0x6ab6d61428fde76768d7b45d8bfeec19c6ef91a8',
            tokenDecimals: 18,
            earnedToken: 'WONE',
            earnedTokenAddress: '0xcf664087a5bb0237a0bad6742852ec6c8d69a27a',
            earnedTokenDecimals: 18,
            earnContractAddress: '0x5b96bbaca98d777cb736dd89a519015315e00d02',
            oracle: 'tokens',
            oracleId: 'BIFI',
            status: 'active',
            retireReason: 'rewards',
            platformId: 'beefy',
            assets: ['BIFI'],
            risks: [
              'COMPLEXITY_LOW',
              'BATTLE_TESTED',
              'IL_NONE',
              'MCAP_SMALL',
              'AUDIT',
              'CONTRACTS_VERIFIED',
            ],
            strategyTypeId: 'single',
            isGovVault: true,
            network: 'harmony',
            createdAt: 1623706144,
          },

          // have one lp vault
          {
            id: 'sushi-one-1ygg-1eth-eol',
            name: '1YGG-1ETH LP',
            token: '1YGG-1ETH LP',
            tokenAddress: '0x40112850EFd90e9e17b56de35d86BdFf9f4d07Fd',
            tokenDecimals: 18,
            tokenProviderId: 'sushi',
            tokenAmmId: 'harmony-sushi',
            earnedToken: 'mooSushiYGG-ETH',
            earnedTokenAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
            earnContractAddress: '0x69F683aa79ab78B95A9823Ace8b9771Ad18fFC71',
            oracle: 'lps',
            oracleId: 'sushi-one-1ygg-1eth',
            status: 'active',
            retireReason: 'rewards',
            platformId: 'sushi',
            assets: ['YGG', 'ETH'],
            risks: [
              'COMPLEXITY_LOW',
              'BATTLE_TESTED',
              'IL_LOW',
              'MCAP_MEDIUM',
              'AUDIT',
              'CONTRACTS_VERIFIED',
            ],
            strategyTypeId: 'lp',
            buyTokenUrl:
              'https://app.sushi.com/swap?inputCurrency=0x6983D1E6DEf3690C4d616b13597A09e6193EA013&outputCurrency=0x63cf309500d8be0B9fDB8F1fb66C821236c0438c',
            addLiquidityUrl:
              'https://www.sushi.com/earn/hmy-s0:0x40112850efd90e9e17b56de35d86bdff9f4d07fd/add',
            removeLiquidityUrl:
              'https://www.sushi.com/earn/hmy-s0:0x40112850efd90e9e17b56de35d86bdff9f4d07fd/remove',
            network: 'harmony',
            createdAt: 1631279238,
          },
        ],
      },
    };
    const action = { type: fetchAllVaults.fulfilled, payload: payload };
    const state = platformsSlice.reducer(initialPlatformsState, action);
    expect(state).toMatchSnapshot();
  });

  it('should update state on fulfilled boosts list', () => {
    const payload: FulfilledAllBoostsPayload = {
      harmony: [
        // one with a partnership
        {
          id: 'moo_banana-pera',
          poolId: 'banana-banana-eol',
          name: 'Pera',
          logo: 'degens/BANANA.svg',
          earnedToken: 'PERA',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xb9D8592E16A9c1a3AE6021CDDb324EaC1Cbc70d6',
          earnContractAddress: '0x38247fCE28480A7BEF2CB7aD134ce091Bd2E1a82',
          earnedOracle: 'tokens',
          earnedOracleId: 'PERA',
          partnership: true,
          status: 'closed',
          isMooStaked: true,
          partners: [
            {
              logo: 'stake/pera/logo.png',
              background: 'stake/pera/bg.png',
              text: "Pera Finance is a decentralized trading competition platform with multi-layered yield farming opportunities. Traders, liquidity providers and holders yield farm together through the DeFi's first decentralized trading competition. Every on-chain PERA transaction (transfer, trade, or liquidity addition/ removal) generates a 2 % transaction fee.",
              website: 'https://pera.finance/',
              social: {
                telegram: 'https://t.me/perafinance',
                twitter: 'https://twitter.com/perafinance',
              },
            },
          ],
        },

        // one closed boost
        {
          id: 'moo_aave-btc',
          poolId: 'aave-wbtc',
          name: 'Iron / Garuda / Fanatics',
          logo: 'single-assets/BTCB.svg',
          earnedToken: 'mooPolygonBIFI',
          earnedTokenDecimals: 18,
          earnedTokenAddress: '0xfEcf784F48125ccb7d8855cdda7C5ED6b5024Cb3',
          earnContractAddress: '0x20948Cad130c3D7B24d27CC66163b4aaed4684F0',
          earnedOracle: 'tokens',
          earnedOracleId: 'BIFI',
          partnership: true,
          status: 'closed',
          isMooStaked: true,
          partners: [],
        },
      ],
    };
    const action = { type: fetchAllBoosts.fulfilled, payload: payload };
    const state = platformsSlice.reducer(initialPlatformsState, action);
    expect(state).toMatchSnapshot();
  });
});
