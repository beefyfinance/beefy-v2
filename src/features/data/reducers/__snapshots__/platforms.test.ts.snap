// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Platforms slice tests should update state on fulfilled boosts list 1`] = `
Object {
  "allIds": Array [
    "beefy.finance",
  ],
  "byChainId": Object {
    "harmony": Array [
      "beefy.finance",
    ],
  },
  "byId": Object {
    "beefy.finance": Object {
      "id": "beefy.finance",
      "name": "Beefy.Finance",
    },
  },
}
`;

exports[`Platforms slice tests should update state on fulfilled vault list 1`] = `
Object {
  "allIds": Array [
    "beefy.finance",
    "sushi",
  ],
  "byChainId": Object {
    "harmony": Array [
      "beefy.finance",
      "sushi",
    ],
  },
  "byId": Object {
    "beefy.finance": Object {
      "id": "beefy.finance",
      "name": "Beefy.Finance",
    },
    "sushi": Object {
      "id": "sushi",
      "name": "Sushi",
    },
  },
}
`;
